File: whisper/pkg/node/whisper_node.go
package node

import (
	"bufio"
	"fmt"
	"net"
	"os"
	"strings"
)

type Node struct {
	conn        net.Conn
	id          string
	subscribers map[string]func(string) // 토픽별 메세지 핸들러
}

// 브로커에 연결
func ConnectToBroker(address string) (*Node, error) {
	conn, err := net.Dial("tcp", address)
	if err != nil {
		return nil, err
	}
	node := &Node{
		conn:        conn,
		id:          conn.LocalAddr().String(),
		subscribers: make(map[string]func(string)),
	}
	return node, nil
}

// 서버에 구독 요청
func (n *Node) Subscribe(topic string) error {
	payload := fmt.Sprintf("SUB %s\n", topic)
	_, err := n.conn.Write([]byte(payload))
	return err
}

// 토픽별 콜백 등록
func (n *Node) RegisterHandler(topic string, handler func(string)) {
	n.subscribers[topic] = handler
}

func handleChat (msg string) {
	fmt.Println("[Chat] Received:", msg)
}

func (n *Node) SetupHandler() {

	n.RegisterHandler("chat", handleChat)
}

// 메세지 송신
func (n *Node) Publish(topic string, message string) error {
	payload := fmt.Sprintf("PUB %s %s\n", topic, message)
	_, err := n.conn.Write([]byte(payload))
	return err
}

func (n *Node) HandleUserInput() {
	scanner := bufio.NewScanner(os.Stdin)
	fmt.Println("Type commands: (e.g., 'SUB topic', 'PUB topic message')")
	for scanner.Scan() {
		input := scanner.Text()
		parts := strings.SplitN(input, " ", 3) // 최대 3개로 분할
		if len(parts) < 2 {
			fmt.Println("Invalid command. Usage: SUB <topic> | PUB <topic> <message>")
			continue
		}
		command := strings.ToUpper(parts[0])
		topic := parts[1]

		switch command {
		case "SUB":
			err := n.Subscribe(topic)
			if err != nil {
				fmt.Println("Subscribe error:", err)
			} else {
				fmt.Println("Subscribed to topic:", topic)
			}
		case "PUB":
			if len(parts) < 3 {
				fmt.Println("Publish needs a topic and a message. Usage: PUB <topic> <message>")
				continue
			}
			message := parts[2]
			err := n.Publish(topic, message)
			if err != nil {
				fmt.Println("Publish error:", err)
			}
		default:
			fmt.Println("Unknown command:", command)
		}
	}
	if err := scanner.Err(); err != nil {
		fmt.Println("Input error:", err)
	}
}

func (n *Node) Listen() {
	buf := make([]byte, 1024)
	for {
		nRead, err := n.conn.Read(buf)
		if err != nil {
			fmt.Println("Read error:", err)
			break
		}
		if nRead > 0 {
			msg := string(buf[:nRead])
			lines := strings.Split(strings.TrimSpace(msg), "\n")
			for _, line := range lines {
				parts := strings.SplitN(line, " ", 3)
				if len(parts) < 3 {
					fmt.Println("Malformed message:", line)
					continue
				}
				command, topic, body := parts[0], parts[1], parts[2]
			
				if command == "MSG" {
					if handler, ok := n.subscribers[topic]; ok {
						handler(body)
					} else {
						fmt.Println("No handler registered for topic:", topic)
					}
				}
			}
		}
	}
}


/*
func main() {
	node, err := ConnectToBroker("localhost:8080")
	if err != nil {
		panic(err)
	}

	// 토픽별 콜백 함수 등록
	node.RegisterHandler("news", func(msg string) {
		fmt.Println("[news] Received:", msg)
	})

	node.RegisterHandler("sports", func(msg string) {
		fmt.Println("[sports] Received:", msg)
	})

	// 브로커에 구독 요청
	node.Subscribe("news")
	node.Subscribe("sports")

	go node.Listen()

	node.Publish("news", "Hello from node!")
	node.Publish("sports", "Soccer match at 8PM")

	select {} // 프로그램 종료 방지
}


*/

File: whisper_rust/src/node/whisper_node.rs
use std::collections::HashMap;
use std::io::{self, BufRead, BufReader, Write};
use std::net::TcpStream;
use std::sync::{Arc, Mutex};
use std::thread;

type Handler = Box<dyn Fn(String) + Send + Sync>;

struct Node {
    conn: Arc<Mutex<TcpStream>>,
    id: String,
    subscribers: Arc<Mutex<HashMap<String, Handler>>>,
}

impl Node {
    fn connect_to_broker(address: &str) -> io::Result<Self> {
        let conn = TcpStream::connect(address)?;
        let id = conn.local_addr()?.to_string();
        Ok(Node {
            conn: Arc::new(Mutex::new(conn)),
            id,
            subscribers: Arc::new(Mutex::new(HashMap::new())),
        })
    }

    fn subscribe(&self, topic: &str) -> io::Result<()> {
        let payload = format!("SUB {}\n", topic);
        self.conn.lock().unwrap().write_all(payload.as_bytes())
    }

    fn register_handler<F>(&self, topic: &str, handler: F)
    where
        F: Fn(String) + Send + Sync + 'static,
    {
        self.subscribers
            .lock()
            .unwrap()
            .insert(topic.to_string(), Box::new(handler));
    }

    fn publish(&self, topic: &str, message: &str) -> io::Result<()> {
        let payload = format!("PUB {} {}\n", topic, message);
        self.conn.lock().unwrap().write_all(payload.as_bytes())
    }

    fn handle_user_input(&self) {
        let node = self.clone();
        thread::spawn(move || {
            let stdin = io::stdin();
            let mut input = String::new();

            println!("Type commands: (e.g., 'SUB topic', 'PUB topic message')");
            loop {
                input.clear();
                if stdin.read_line(&mut input).unwrap() == 0 {
                    break;
                }
                let parts: Vec<&str> = input.trim().splitn(3, ' ').collect();
                if parts.len() < 2 {
                    println!("Invalid command. Usage: SUB <topic> | PUB <topic> <message>");
                    continue;
                }
                let command = parts[0].to_uppercase();
                let topic = parts[1];

                match command.as_str() {
                    "SUB" => {
                        if let Err(e) = node.subscribe(topic) {
                            println!("Subscribe error: {}", e);
                        } else {
                            println!("Subscribed to topic: {}", topic);
                        }
                    }
                    "PUB" => {
                        if parts.len() < 3 {
                            println!("Publish needs a topic and a message. Usage: PUB <topic> <message>");
                            continue;
                        }
                        let message = parts[2];
                        if let Err(e) = node.publish(topic, message) {
                            println!("Publish error: {}", e);
                        }
                    }
                    _ => println!("Unknown command: {}", command),
                }
            }
        });
    }

    fn listen(&self) {
        let conn = self.conn.lock().unwrap().try_clone().unwrap();
        let subscribers = Arc::clone(&self.subscribers);

        thread::spawn(move || {
            let reader = BufReader::new(conn);
            for line in reader.lines() {
                match line {
                    Ok(msg) => {
                        let parts: Vec<&str> = msg.trim().splitn(3, ' ').collect();
                        if parts.len() < 3 {
                            println!("Malformed message: {}", msg);
                            continue;
                        }
                        let command = parts[0];
                        let topic = parts[1];
                        let body = parts[2];

                        println!("Raw Message : {}", body);

                        if command == "MSG" {
                            if let Some(handler) = subscribers.lock().unwrap().get(topic) {
                                handler(body.to_string());
                            } else {
                                println!("No handler registered for topic: {}", topic);
                            }
                        }
                    }
                    Err(e) => {
                        println!("Read error: {}", e);
                        break;
                    }
                }
            }
        });
    }
}

impl Clone for Node {
    fn clone(&self) -> Self {
        Node {
            conn: Arc::clone(&self.conn),
            id: self.id.clone(),
            subscribers: Arc::clone(&self.subscribers),
        }
    }
}

fn main() -> io::Result<()> {
    let node = Node::connect_to_broker("127.0.0.1:8080")?;

    // 예시: 토픽별 핸들러 등록
    node.register_handler("chat", |msg| {
        println!("[Chat] Received: {}", msg);
    });

    // 브로커에 구독 요청
    node.subscribe("chat")?;

    // 서버 수신 대기
    node.listen();

    // 사용자 입력 처리
    node.handle_user_input();

    // 프로그램 종료 방지
    loop {
        thread::park();
    }
}
